solr/core/src/java/org/apache/solr/update/processor/DistributedUpdateProcessor.java

```diff
@@ -331,7 +342,7 @@ protected boolean versionAdd(AddUpdateCommand cmd) throws IOException {
 
     boolean isReplayOrPeersync =
         (cmd.getFlags() & (UpdateCommand.REPLAY | UpdateCommand.PEER_SYNC)) != 0;
-    boolean leaderLogic = isLeader && !isReplayOrPeersync;
+    boolean leaderLogic = leaderLogicWithVersionIntegrityCheck(isReplayOrPeersync, versionOnUpdate);
```

Comment:
I don't like this change (thus don't like leaderLogicWithVersionIntegrityCheck).  If there is some race condition causing the client to route to us (a new sub-shard) before we know that we are ready to accept it, I'd rather not pretend we are ready in any way.  Maybe throw an exception like INVALID_STATE that should cause CloudSolrClient to retry.  Our state will update eventually.

Reply:
My concern is better expressed as related to aspects of the contents of leaderLogicWithVersionIntegrityCheck().  I like that it throws an exception when versionOnUpdate = 0 when we expect it to not be 0.  But I don't like changing the criteria for "leaderLogic".

Reply:
I don't see ShardSplitTest _indexing_ docs with an HttpClient.  Except for the "controlClient" maybe, which probably doesn't count but I didn't look closer at that aspect.

Reply:
I am not as confident about the above anymore looking at the error traces. not sure how, but the test is using CloudSolrClient for some of the requests, I am seeing the following in the logs
```
  2> 11376 INFO  (Thread-43) [] o.a.s.c.s.i.CloudSolrClient Request to collection [collection1] failed due to (400) org.apache.solr.client.solrj.impl.BaseHttpSolrClient$RemoteSolrException: Error from server at http://127.0.0.1:59362/collection1: missing _version_ on update from leader, retry=0 maxRetries=5 commError=false errorCode=400 
```

I am considering changing the response code from BAD_REQUEST to INVALID_STATE to allow for a retry (same as #1484). thanks for the hint @dsmiley, it didn't click when you mentioned it earlier (what was missing was the extra check).